airbyte-ci/connectors/metadata_service/lib/metadata_service/registry.py

```diff
@@ -0,0 +1,373 @@
+    """
+    connector_id = registry_entry_dict[ConnectorTypePrimaryKey[connector_type.value]]
+    metrics = latest_metrics_dict.get(connector_id, {})
+
+    # Safely add metrics to ["generated"]["metrics"], knowing that the key may not exist, or might be None
+    registry_entry_dict = set_with(registry_entry_dict, "generated.metrics", metrics, default_none_to_dict)
+
+    return registry_entry_dict
+
+
+def _apply_release_candidate_entries(registry_entry_dict: dict, docker_repository_to_rc_registry_entry: dict) -> dict:
+    """Apply the optionally existing release candidate entries to the registry entry.
+    We need both the release candidate metadata entry and the release candidate registry entry because the metadata entry contains the rollout configuration, and the registry entry contains the actual RC registry entry.
+
+    Args:
+        registry_entry_dict (dict): The registry entry.
+        docker_repository_to_rc_registry_entry (dict): Mapping of docker repository to release candidate registry entry.
+
+    Returns:
+        dict: The registry entry with release candidates applied.
+    """
+    registry_entry_dict = copy.deepcopy(registry_entry_dict)
+    if registry_entry_dict["dockerRepository"] in docker_repository_to_rc_registry_entry:
+        release_candidate_registry_entry = docker_repository_to_rc_registry_entry[registry_entry_dict["dockerRepository"]]
+        registry_entry_dict = _apply_release_candidates(registry_entry_dict, release_candidate_registry_entry)
+    return registry_entry_dict
+
+
+def _apply_release_candidates(
+    latest_registry_entry: dict,
+    release_candidate_registry_entry: PolymorphicRegistryEntry,
+) -> dict:
+    """Apply the release candidate entries to the registry entry.
+
+    Args:
+        latest_registry_entry (dict): The latest registry entry.
+        release_candidate_registry_entry (PolymorphicRegistryEntry): The release candidate registry entry.
+
+    Returns:
+        dict: The registry entry with release candidates applied.
+    """
+    try:
+        if not release_candidate_registry_entry.releases.rolloutConfiguration.enableProgressiveRollout:
+            return latest_registry_entry
+    # Handle if releases or rolloutConfiguration is not present in the release candidate registry entry
+    except AttributeError:
+        return latest_registry_entry
+
+    # Ensure that the release candidate is newer than the latest registry entry
+    if semver.Version.parse(release_candidate_registry_entry.dockerImageTag) < semver.Version.parse(
+        latest_registry_entry["dockerImageTag"]
+    ):
+        return latest_registry_entry
+
+    updated_registry_entry = copy.deepcopy(latest_registry_entry)
+    updated_registry_entry.setdefault("releases", {})
+    updated_registry_entry["releases"]["releaseCandidates"] = {
+        release_candidate_registry_entry.dockerImageTag: to_json_sanitized_dict(release_candidate_registry_entry)
+    }
```

Comment:
for my understanding: we generate a struct like this
```yaml
releases:
  releaseCandidates:
    1.2.3-rc.1: {...}
```
and _remove_ any previous RCs from that `releaseCandidates` dict?